File: android/src/main/kotlin/com/kurmantaev/vk_id_auth/VkIdAuthPluginExtensions.kt
package com.kurmantaev.vk_id_auth

import com.vk.id.AccessToken
import com.vk.id.VKIDUser
import io.flutter.plugin.common.MethodChannel

/// Расширение над классом [AccessToken] VK ID SDK для конвертации объекта [AccessToken]
/// в [Map<String, Any?>].
val AccessToken.toMap: Map<String, Any?>
    get() = mapOf<String, Any?>(
        "token" to token,
        "userId" to userID,
        "userData" to userData.toMap,
    )

/// Расширение над классом [VKIDUser] VK ID SDK для конвертации объекта [VKIDUser] в
/// [Map<String, Any?>].
val VKIDUser.toMap: Map<String, Any?>
    get() = mapOf<String, Any?>(
        "firstName" to firstName,
        "lastName" to lastName,
        "email" to email,
        "phone" to phone
    )

/// Отправляет [VkIdAuthPluginError] в Flutter.
fun MethodChannel.Result.pluginErrorToFlutter(error: VkIdAuthPluginError) = error(
    error.code,
    error.message,
    error.details
)

File: ios/Classes/VkIdAuthPluginExtensions.swift
import VKID

/// Расширение над классом [UserSession] VK ID SDK для конвертации объекта [UserSession]
/// в [Dictionary<String, Any?>].
extension UserSession {

    var toDictionary: Dictionary<String, Any?> {
        var dictionary = Dictionary<String, Any?>()
        let token = accessToken.value
        let userId = user?.id.value
        let user = user?.toDictionary
        dictionary.updateValue(token, forKey: "token")
        dictionary.updateValue(userId, forKey: "userId")
        dictionary.updateValue(user, forKey: "userData")
        return dictionary
    }

}

/// Расширение над классом [User] VK ID SDK для конвертации объекта [User]
/// в [Dictionary<String, Any?>].
extension User {

    var toDictionary: Dictionary<String, Any?> {
        var dictionary = Dictionary<String, Any?>()
        dictionary.updateValue(firstName, forKey: "firstName")
        dictionary.updateValue(lastName, forKey: "lastName")
        dictionary.updateValue(email, forKey: "email")
        dictionary.updateValue(phone, forKey: "phone")
        return dictionary
    }
}
